apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java

```diff
@@ -128,6 +128,9 @@
 
     private final Map<String, Object> otelAttributes = new HashMap<>();
 
+    // TODO properly size to limit allocation
+    private final Map<Class<?>, ElasticContext<?>> contextWrappers = new HashMap<>();
```

Comment:
The same span can be active in multiple threads at the same time. The map of context wrappers needs to be maintained by the `activeStack` in `ElasticApmTracer`. Maybe on the first time an OTel context is requested, upgrade the `Span` `ElasticContext` to a `ElasticContext` that has a context wrapper map.

API wise, it could look like `tracer.getActive(OTelBridgeContext.class, span -> OTelBridgeContext.wrapElasticActiveSpan(tracer, currentSpan))`.

Reply:
The wrapped span won't be used for activation/deactivation (and we could even throw an exception if we ever try to), so in theory we should not have issues with creating or overwriting more than once for the same span (we might have to switch to a concurrent-safe collection though).

Reply:
Consider the following scenario:

Thread 1 activates context#1 with context objects: span:span A, foo:42
Thread 2 activates context#2 with context objects: span:span A, foo:21
Thread 1 gets active context
-- expected: context#1
-- actual: context#2

The activation of context#2 overrides context#1 in the Span#contextWrappers map.
As a span can be active in multiple contexts, it's not the proper place to store the context wrapper objects.

Reply:
Thanks for the very clear example here. While the context objects do not hold any state and should *in theory* be interchangeable and strictly equivalent as long as they belong to the same classloader, this would break if there is any condition or assumption using reference equality, which seems to be a common usage pattern with OTel, hence making it thread-local would avoid this potential issue.

Reply:
> While the context objects do not hold any state 

Context objects do hold state. In my example, they contain the same span and a `ContextKey` named `foo`. Thread 1 would have expected the value `42` for the `foo` context object but it only gets half of the answer (`21`).

> making it thread-local would avoid this potential issue.

The thread local map should, however, not be in the Span object. We should only have a single place where we store thread-local state: `co.elastic.apm.agent.impl.ElasticApmTracer#activeStack`.